From:	Stefano Stabellini <stefano.stabellini@eu.citrix.com>
Subject: x86: Save cr4 to mmu_cr4_features at boot time
References: bnc#684248
Patch-mainline: submitted 31 Mar

Save cr4 to mmu_cr4_features at boot time

This patch fixes a freeze when resuming from hibernation, introduced by
"x86: Cleanup highmap after brk is concluded", commit id
e5f15b45ddf3afa2bbbb10c7ea34fb32b6de0a0e.
This patch should be backported to all the stable tree where the
offending commit is present.

Signed-off-by: Stefano Stabellini <stefano.stabellini@eu.citrix.com>
Tested-by: Michael Leun <lkml20101129@newton.leun.net>
CC: Yinghai Lu <yinghai@kernel.org>
CC: stable <stable@kernel.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>

diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
index 5a0484a..0943eb2 100644
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -891,6 +891,7 @@ void __init setup_arch(char **cmdline_p)
 		max_low_pfn = max_pfn;
 
 	high_memory = (void *)__va(max_pfn * PAGE_SIZE - 1) + 1;
+	mmu_cr4_features = read_cr4();
 #endif
 
 	/*
